docs/reference/configuration.md
# Configuration

Predictive Horizontal Pod Autoscalers have a number of configuration options available.

## minReplicas

```yaml
minReplicas: 2
```

The lower limit for the number of replicas to which the autoscaler can scale down to. `minReplicas` is allowed to be 0
if at least one Object or External metric is configured. Scaling is active as long as at least one metric value is
available.

Default value: `1`.

## maxReplicas

```yaml
maxReplicas: 15
```

The upper limit for the number of replicas to which the autoscaler can scale up.
It cannot be less than minReplicas.

Default value: `10`.

## syncPeriod

```yaml
syncPeriod: 10000
```

Equivalent to `--horizontal-pod-autoscaler-sync-period`; the frequency with which the PHPA calculates replica counts and
scales in milliseconds.

Set in milliseconds.

Default value: `15000` (15 seconds).

Set in milliseconds.

## cpuInitializationPeriod

```yaml
cpuInitializationPeriod: 150
```

Equivalent to `--horizontal-pod-autoscaler-cpu-initialization-period`; the period after pod start when CPU samples
might be skipped.

Set in seconds.

Default value: `300` (5 minutes).

## initialReadinessDelay

```yaml
initialReadinessDelay: 45
```

Equivalent to `--horizontal-pod-autoscaler-initial-readiness-delay`; the period after pod start during which readiness
changes will be treated as initial readiness.

Set in seconds.

Default value: `30` (30 seconds).

## tolerance

```yaml
tolerance: 0.25
```

Equivalent to `--horizontal-pod-autoscaler-tolerance`; the minimum change (from 1.0) in the desired-to-actual metrics
ratio for the horizontal pod autoscaler to consider scaling.

Default value: `0.1`.

## decisionType

```yaml
decisionType: mean
```

Decider on which evaluation to pick if there are multiple models provided.

Possible values:

- **maximum** - pick the highest evaluation of the models.
- **minimum** - pick the lowest evaluation of the models.
- **mean** - calculate the mean number of replicas (rounded to nearest integer) between the models.
- **median** - calculate the median number of replicas between the models.

Default value: `maximum`.

## behavior

Scaling behavior to apply.

Intended to be feature equivalent to Kubernetes HPA behavior.

See the [Horizontal Pod Autoscaler docs
here](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#configurable-scaling-behavior).

## models

List of statistical models to apply.
See [the models section for details](../../user-guide/models).

## metrics

List of metrics to target for evaluating replica counts.
See [the metrics section for details](../../user-guide/metrics).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic
Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [v0.13.0] - 2023-01-16
### Added
- New ability to choose a start interval for a model, e.g. a holt winters model that only starts at the next full hour.
  - `startInterval` is an optional [duration](https://pkg.go.dev/time#ParseDuration) that defines the next interval to
  start at. For example `1m` would start the model recording data and calculating after the next full minute has passed.
- New ability to clear a model's replica history if too much time has passed since it last recorded any data.
  - `resetDuration` is an optional [duration](https://pkg.go.dev/time#ParseDuration) that defines how long should be
  allowed to pass without a model recording any data, if that occurs the replica history is cleared and if a start
  interval is provided a new start time is calculated. If a model is not calculated for an extended period of time
  (e.g. a cluster being powered off) this allows old data to be cleared out and not used in calculations and a new
  start time calculated to respect any provided interval. For example `3m` would clear the model's replica history
  (and reset the start time if interval is provided) if it has been more than 3 minutes since the last data was
  recorded for the model.
### Changed
- **BREAKING CHANGE** typo fixed, `ModelHistory.ReplicaHistroy` renamed to `ModelHistory.ReplicaHistory`.

## [v0.12.0] - 2023-01-15
### Changed
- See the [migration guide from `v0.11.2`
here](https://predictive-horizontal-pod-autoscaler.readthedocs.io/en/latest/user-guide/migration/v0_11_2-to-v0_12_0).
- **BREAKING CHANGE** PHPA spec upgraded from `autoscaling/v2beta2` to `autoscaling/v2` for the following definitions:
  - `CrossVersionObjectReference` in the `scaleTargetRef` field.
  - `MetricSpec` in the `metrics` field.
  - `MetricStatus` in the `currentMetrics` field.
- Upgraded to [k8shorizmetrics `v2.0.0`](https://github.com/jthomperoo/k8shorizmetrics/releases/tag/v2.0.0).
- Upgraded from `autoscaling/v2beta2` to `autoscaling/v2`.
- Upgraded to Go `v1.19`.
### Removed
- **BREAKING CHANGE** Removed `downscaleStabilization`, replaced with `behavior`, `scaleDown`,
`stabilizationWindowSeconds`.
### Added
- Support for [HPA scaling
behaviors](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#configurable-scaling-behavior).

## [v0.11.2] - 2022-11-25
### Fixed
- Bug that would display a statsmodels error when using Holt-Winters and having too few observations (less than 2 *
seasonal periods).

## [v0.11.1] - 2022-07-23
### Fixed
- Bug that caused the autoscaler to fail to scale every time, due to invalid K8s GVR resource fed to the scaling
client.

## [v0.11.0] - 2022-07-23
### Changed
- See the [migration guide from `v0.10.0` here](https://predictive-horizontal-pod-autoscaler.readthedocs.io/en/latest/user-guide/migration/v0_10_0-to-v0_11_0).
- BREAKING CHANGE: Major rewrite converting this project from a Custom Pod Autoscaler to have its own dedicated CRD
and operator.
  - Configuration and deployment has changed completely, no longer need to install the Custom Pod Autoscaler Operator,
  instead you need to install the Predictive Horizontal Pod Autoscaler as an operator.
  - No longer deployed as `CustomPodAutoscaler` custom resources, now deployed as `PredictiveHorizontalPodAutoscaler`
  custom resources.
- BREAKING CHANGES: Several configuration options renamed for clarity.
  - `LinearModel -> storedValues` renamed to `LinearModel -> historySize`
  - `Model -> perInterval` renamed to `Model -> perSyncPeriod`
- BREAKING CHANGES: `perSyncPeriod` behaviour changed slightly, now it will no longer calculate the prediction, but
it will still update the replica history available to make a prediction with and prune the replica history if needed.
- Holt-Winters runtime hooks format changed:
  - `evaluations` renamed to `replicaHistory`.
  - Format change for `replicaHistory`, now in the format `[{"time": "<timestamp>", "replicas": <replica count>}]`.
- Upgrade to Go `v1.18`.
- No longer SQLite based storage, instead using Kubernetes configmaps which give persistent storage by default with
resiliency when the autoscaler operator restarts.
### Removed
- BREAKING CHANGES: Removed some no longer relevant configuration options.
  - `DBPath`
  - `MigrationPath`
- BREAKING CHANGE: Since no longer built as a `CustomPodAutoscaler` the `startTime` configuration is no longer
available: <https://custom-pod-autoscaler.readthedocs.io/en/latest/reference/configuration/#starttime>.
- BREAKING CHANGE: Holt-Winters runtime hooks limited to only support HTTP, shell support dropped.

## [v0.10.0] - 2022-05-14
### Changed
- Removed dependency on `jthomperoo/horizontal-pod-autoscaler` in favour of `jthomperoo/k8shorizmetrics`.
- Bump `jthomperoo/custom-pod-autoscaler` to `v2.6.0`.
- Upgrade to Go `v1.17`.

## [v0.9.0] - 2021-12-28
### Added
- Support for `argoproj.io/v1alpha1` `Rollout` resource.
### Changed
- Bump `jthomperoo/custom-pod-autoscaler` to `v2.3.0`
- Bump `jthomperoo/horizontal-pod-autoscaler` to `v0.8.0`

## [v0.8.0] - 2021-08-15
### Changed
- Updated to Custom Pod Autoscaler `v2.2.0`.
- Updated to Horizontal Pod Autoscaler CPA `v0.7.0`.
### Fixed
- Linear regression no longer fails on first run.

## [v0.7.0] - 2021-04-05
### Added
- Holt Winters values can now be fetched at runtime, rather than simply being hardcoded.
### Fixed
- Fixed slow shutdown of PHPA due to ignoring SIGTERM from K8s.
### Changed
- Switched from Golang to Python for calculating statistical predictions for Linear Regression and Holt-Winters.
- Holt-Winters now calculated using statsmodels, opening up statsmodels configuration options for tuning.

internal/config/config.go
/*
Copyright 2021 The Predictive Horizontal Pod Autoscaler Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Package config provides configuration options for the Predictive Horizontal Pod Autoscaler
package config

import (
	"io"

	"github.com/jthomperoo/predictive-horizontal-pod-autoscaler/internal/hook"
	autoscalingv2 "k8s.io/api/autoscaling/v2beta2"
	"k8s.io/apimachinery/pkg/util/yaml"
)

const (
	// DecisionMaximum means use the highest predicted value from the models
	DecisionMaximum = "maximum"
	// DecisionMinimum means use the lowest predicted value from the models
	DecisionMinimum = "minimum"
	// DecisionMean means use the mean average of predicted values
	DecisionMean = "mean"
	// DecisionMedian means use the median average of predicted values
	DecisionMedian = "median"
)

const (
	defaultTolerance = float64(0.1)
	// 5 minute CPU initialization period
	defaultCPUInitializationPeriod = 300
	// 30 second initial readiness delay
	defaultInitialReadinessDelay = 30
)

// Config holds the configuration of the Predictive element of the PHPA
type Config struct {
	Models                  []*Model                   `json:"models"`
	Metrics                 []autoscalingv2.MetricSpec `json:"metrics"`
	DecisionType            string                     `json:"decisionType"`
	DBPath                  string                     `json:"dbPath"`
	MigrationPath           string                     `json:"migrationPath"`
	Tolerance               float64                    `json:"tolerance"`
	CPUInitializationPeriod int                        `json:"cpuInitializationPeriod"`
	InitialReadinessDelay   int                        `json:"initialReadinessDelay"`
}

// Model represents a prediction model to use, e.g. a linear regression
type Model struct {
	Type               string       `json:"type"`
	Name               string       `json:"name"`
	PerInterval        int          `json:"perInterval"`
	CalculationTimeout *int         `json:"calculationTimeout"`
	Linear             *Linear      `json:"linear"`
	HoltWinters        *HoltWinters `json:"holtWinters"`
}

// HoltWinters represents a holt-winters exponential smoothing prediction model configuration
type HoltWinters struct {
	Alpha                  *float64         `json:"alpha"`
	Beta                   *float64         `json:"beta"`
	Gamma                  *float64         `json:"gamma"`
	Trend                  string           `json:"trend"`
	Seasonal               string           `json:"seasonal"`
	SeasonalPeriods        int              `json:"seasonalPeriods"`
	StoredSeasons          int              `json:"storedSeasons"`
	DampedTrend            *bool            `json:"dampedTrend"`
	InitializationMethod   *string          `json:"initializationMethod"`
	InitialLevel           *float64         `json:"initialLevel"`
	InitialTrend           *float64         `json:"initialTrend"`
	InitialSeasonal        *float64         `json:"initialSeasonal"`
	RuntimeTuningFetchHook *hook.Definition `json:"runtimeTuningFetchHook"`
}

// Linear represents a linear regression prediction model configuration
type Linear struct {
	StoredValues int `json:"storedValues"`
	LookAhead    int `json:"lookAhead"`
}

// LoadConfig takes in the predictive config as a byte array and uses it to build the config, overriding default values
func LoadConfig(configEnv io.Reader) (*Config, error) {
	config := newDefaultConfig()
	err := yaml.NewYAMLOrJSONDecoder(configEnv, 10).Decode(&config)
	if err != nil {
		return nil, err
	}
	return config, nil
}

func newDefaultConfig() *Config {
	return &Config{
		DecisionType:            "maximum",
		DBPath:                  "/store/predictive-horizontal-pod-autoscaler.db",
		MigrationPath:           "/app/sql",
		Tolerance:               defaultTolerance,
		CPUInitializationPeriod: defaultCPUInitializationPeriod,
		InitialReadinessDelay:   defaultInitialReadinessDelay,
	}
}
